File: LeetCode-In-Java/src/BreadthFirstSearch/LeetCode_107.java
package BreadthFirstSearch;

import java.util.*;

/**
 * TODO
 *
 * @author <a href="mailto:ge.sf.chn@gmail.com">shaofeng</a>
 * @see
 * @since 2019/5/15
 */
public class LeetCode_107 {

    public List<List<Integer>> levelOrderBottom(TreeNode root) {
        Queue<TreeNode> queue = new LinkedList<>();
        queue.add(root);
        List<List<Integer>> lists = new ArrayList<>();
        int size;
        while (root != null && (size = queue.size()) > 0) {
            List<Integer> list = new ArrayList<>();
            for (int i = 0; i < size; i++) {
                TreeNode cur = queue.poll();
                if (cur.left != null) {
                    queue.add(cur.left);
                }
                if (cur.right != null) {
                    queue.add(cur.right);
                }
                list.add(cur.val);
            }
            lists.add(list);
        }

        Collections.reverse(lists);
        return lists;
    }
}

File: LeetCode-In-Go/internal/bfs/LeetCode_107.go
package bfs

func levelOrderBottom(root *TreeNode) [][]int {
	res := [][]int{}
	if root == nil {
		return res
	}

	arr := append([]TreeNode{}, *root)
	for len(arr) > 0 {
		res1 := []int{}
		for _, n := range arr {
			arr = arr[1:]
			if n.Left != nil {
				arr = append(arr, *n.Left)
			}
			if n.Right != nil {
				arr = append(arr, *n.Right)
			}
			res1 = append(res1, n.Val)
		}
		res = append(res, res1)
	}

	reverse(res, 0, len(res)-1)
	return res
}

func reverse(res [][]int, i, j int) {
	for i < j {
		res[i], res[j] = res[j], res[i]
		i++
		j--
	}
}
